### PR TITLE
Handle readBatch changes for Spark 3.3.0

### LINKED ISSUE #5257: [FEA][Audit] - [SPARK-34863][SQL] Support complex types for Parquet vectorized reader
**Is your feature request related to a problem? Please describe.**
 Support for complex types (list, map, array) for Spark's vectorized Parquet reader is added in Spark-3.3. From the PR description, the benchmark shows good speed up when vectorized parquet reader is used for complex types. Need to evaluate the tasks to be done in the plugin to support this. 

Spark commit - https://github.com/apache/spark/commit/deac8f950e

### LINKED ISSUE #5357: [BUG] Spark 3.3 cache_test test_passing_gpuExpr_as_Expr[failures
**Describe the bug**
07:55:49  FAILED ../../src/main/python/cache_test.py::test_passing_gpuExpr_as_Expr[{'spark.sql.inMemoryColumnarStorage.enableVectorizedReader': 'true'}][ALLOW_NON_GPU(CollectLimitExec)]
07:55:49  FAILED ../../src/main/python/cache_test.py::test_passing_gpuExpr_as_Expr[{'spark.sql.inMemoryColumnarStorage.enableVectorizedReader': 'false'}][ALLOW_NON_GPU(CollectLimitExec)]

### LINKED ISSUE #5429: [BUG] test_cache_expand_exec fails on Spark 3.3
Last night's Spark 3.3 test run failed with errors in test_cache_expand_exec.  For example:
```
07:37:48  _ test_cache_expand_exec[{'spark.sql.inMemoryColumnarStorage.enableVectorizedReader': 'false'}-Decimal(12,2)] _
07:37:48  
07:37:48  data_gen = Decimal(12,2)
07:37:48  enable_vectorized_conf = {'spark.sql.inMemoryColumnarStorage.enableVectorizedReader': 'false'}
07:37:48  
07:37:48      @pytest.mark.parametrize('data_gen', all_gen, ids=idfn)
07:37:48      @pytest.mark.parametrize('enable_vectorized_conf', enable_vectorized_confs, ids=idfn)
07:37:48      @ignore_order
07:37:48      def test_cache_expand_exec(data_gen, enable_vectorized_conf):
07:37:48          def op_df(spark, length=2048, seed=0):
07:37:48              cached = gen_df(spark, StructGen([
07:37:48                  ('a', data_gen),
07:37:48                  ('b', IntegerGen())], nullable=False), length=length, seed=seed).cache()
07:37:48              cached.count() # populate the cache
07:37:48              return cached.rollup(f.col("a"), f.col("b")).agg(f.col("b"))
07:37:48      
07:37:48  >       assert_gpu_and_cpu_are_equal_collect(op_df, conf=enable_vectorized_conf)
07:37:48  
07:37:48  ../../src/main/python/cache_test.py:101: 
07:37:48  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
07:37:48  ../../src/main/python/asserts.py:508: in assert_gpu_and_cpu_are_equal_collect
07:37:48      _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
07:37:48  ../../src/main/python/asserts.py:427: in _assert_gpu_and_cpu_are_equal
07:37:48      run_on_cpu()
07:37:48  ../../src/main/python/asserts.py:413: in run_on_cpu
07:37:48      from_cpu = with_cpu_session(bring_back, conf=conf)
07:37:48  ../../src/main/python/spark_session.py:114: in with_cpu_session
07:37:48      return with_spark_session(func, conf=copy)
07:37:48  ../../src/main/python/spark_session.py:98: in with_spark_session
07:37:48      ret = func(_spark)
07:37:48  ../../src/main/python/asserts.py:201: in <lambda>
07:37:48      bring_back = lambda spark: limit_func(spark).collect()
07:37:48  ../../src/main/python/asserts.py:184: in with_sorted
07:37:48      df = func(spark)
07:37:48  ../../src/main/python/cache_test.py:98: in op_df
07:37:48      cached.count() # populate the cache
07:37:48  /home/jenkins/agent/workspace/jenkins-rapids_it-3.3.x-SNAPSHOT-dev-github-80/jars/spark-3.3.0-SNAPSHOT-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/dataframe.py:804: in count
07:37:48      return int(self._jdf.count())
07:37:48  /home/jenkins/agent/workspace/jenkins-rapids_it-3.3.x-SNAPSHOT-dev-github-80/jars/spark-3.3.0-SNAPSHOT-bin-hadoop3.2/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py:1321: in __call__
07:37:48      return_value = get_return_value(
07:37:48  /home/jenkins/agent/workspace/jenkins-rapids_it-3.3.x-SNAPSHOT-dev-github-80/jars/spark-3.3.0-SNAPSHOT-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/utils.py:190: in deco
07:37:48      return f(*a, **kw)
07:37:48  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
07:37:48  
07:37:48  answer = 'xro2660'
07:37:48  gateway_client = <py4j.clientserver.JavaClient object at 0x7f2e05588220>
07:37:48  target_id = 'o2658', name = 'count'
07:37:48  
07:37:48      def get_return_value(answer, gateway_client, target_id=None, name=None):
07:37:48          """Converts an answer received from the Java gateway into a Python object.
07:37:48      
07:37:48          For example, string representation of integers are converted to Python
07:37:48          integer, string representation of objects are converted to JavaObject
07:37:48          instances, etc.
07:37:48      
07:37:48          :param answer: the string returned by the Java gateway
07:37:48          :param gateway_client: the gateway client used to communicate with the Java
07:37:48              Gateway. Only necessary if the answer is a reference (e.g., object,
07:37:48              list, map)
07:37:48          :param target_id: the name of the object from which the answer comes from
07:37:48              (e.g., *object1* in `object1.hello()`). Optional.
07:37:48          :param name: the name of the member from which the answer comes from
07:37:48              (e.g., *hello* in `object1.hello()`). Optional.
07:37:48          """
07:37:48          if is_error(answer)[0]:
07:37:48              if len(answer) > 1:
07:37:48                  type = answer[1]
07:37:48                  value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
07:37:48                  if answer[1] == REFERENCE_TYPE:
07:37:48  >                   raise Py4JJavaError(
07:37:48                          "An error occurred while calling {0}{1}{2}.\n".
07:37:48                          format(target_id, ".", name), value)
07:37:48  E                   py4j.protocol.Py4JJavaError: An error occurred while calling o2658.count.
07:37:48  E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 48.0 failed 1 times, most recent failure: Lost task 0.0 in stage 48.0 (TID 48) (10.233.91.240 executor 0): java.lang.NoSuchMethodException: org.apache.spark.sql.execution.datasources.parquet.VectorizedColumnReader.readBatch(int, org.apache.spark.sql.execution.vectorized.WritableColumnVector)
07:37:48  E                   	at java.lang.Class.getDeclaredMethod(Class.java:2130)
07:37:48  E                   	at com.nvidia.spark.rapids.ParquetCachedBatchSerializer$CachedBatchIteratorConsumer.<init>(ParquetCachedBatchSerializer.scala:693)
07:37:48  E                   	at com.nvidia.spark.rapids.ParquetCachedBatchSerializer.$anonfun$convertCachedBatchToInternalRow$1(ParquetCachedBatchSerializer.scala:595)
07:37:48  E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:855)
07:37:48  E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:855)
07:37:48  E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
07:37:48  E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
07:37:48  E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
07:37:48  E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
07:37:48  E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
07:37:48  E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
07:37:48  E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
07:37:48  E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
07:37:48  E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
07:37:48  E                   	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
07:37:48  E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
07:37:48  E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
07:37:48  E                   	at org.apache.spark.scheduler.Task.run(Task.scala:136)
07:37:48  E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
07:37:48  E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
07:37:48  E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
07:37:48  E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
07:37:48  E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
07:37:48  E                   	at java.lang.Thread.run(Thread.java:748)
07:37:48  E                   
07:37:48  E                   Driver stacktrace:
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2672)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2608)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2607)
07:37:48  E                   	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
07:37:48  E                   	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
07:37:48  E                   	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2607)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1182)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1182)
07:37:48  E                   	at scala.Option.foreach(Option.scala:407)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1182)
07:37:48  E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2860)
07:37:48  E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2802)
07:37:48  E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2791)
07:37:48  E                   	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
07:37:48  E                   	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:952)
07:37:48  E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2228)
07:37:48  E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2249)
07:37:48  E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2268)
07:37:48  E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2293)
07:37:48  E                   	at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1021)
07:37:48  E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
07:37:48  E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
07:37:48  E                   	at org.apache.spark.rdd.RDD.withScope(RDD.scala:406)
07:37:48  E                   	at org.apache.spark.rdd.RDD.collect(RDD.scala:1020)
07:37:48  E                   	at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:424)
07:37:48  E                   	at org.apache.spark.sql.Dataset.$anonfun$count$1(Dataset.scala:3161)
07:37:48  E                   	at org.apache.spark.sql.Dataset.$anonfun$count$1$adapted(Dataset.scala:3160)
07:37:48  E                   	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3856)
07:37:48  E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:109)
07:37:48  E                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
07:37:48  E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
07:37:48  E                   	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
07:37:48  E                   	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
07:37:48  E                   	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3854)
07:37:48  E                   	at org.apache.spark.sql.Dataset.count(Dataset.scala:3160)
07:37:48  E                   	at sun.reflect.GeneratedMethodAccessor77.invoke(Unknown Source)
07:37:48  E                   	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
07:37:48  E                   	at java.lang.reflect.Method.invoke(Method.java:498)
07:37:48  E                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
07:37:48  E                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
07:37:48  E                   	at py4j.Gateway.invoke(Gateway.java:282)
07:37:48  E                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
07:37:48  E                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)
07:37:48  E                   	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
07:37:48  E                   	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
07:37:48  E                   	at java.lang.Thread.run(Thread.java:748)
07:37:48  E                   Caused by: java.lang.NoSuchMethodException: org.apache.spark.sql.execution.datasources.parquet.VectorizedColumnReader.readBatch(int, org.apache.spark.sql.execution.vectorized.WritableColumnVector)
07:37:48  E                   	at java.lang.Class.getDeclaredMethod(Class.java:2130)
07:37:48  E                   	at com.nvidia.spark.rapids.ParquetCachedBatchSerializer$CachedBatchIteratorConsumer.<init>(ParquetCachedBatchSerializer.scala:693)
07:37:48  E                   	at com.nvidia.spark.rapids.ParquetCachedBatchSerializer.$anonfun$convertCachedBatchToInternalRow$1(ParquetCachedBatchSerializer.scala:595)
07:37:48  E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:855)
07:37:48  E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:855)
07:37:48  E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
07:37:48  E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
07:37:48  E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
07:37:48  E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
07:37:48  E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
07:37:48  E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
07:37:48  E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
07:37:48  E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
07:37:48  E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
07:37:48  E                   	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
07:37:48  E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
07:37:48  E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
07:37:48  E                   	at org.apache.spark.scheduler.Task.run(Task.scala:136)
07:37:48  E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
07:37:48  E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
07:37:48  E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
07:37:48  E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
07:37:48  E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
07:37:48  E                   	... 1 more
```

### PR DESCRIPTION
**Pull request by razajafri**
The existing class `CurrentBatchIterator` has been changed to a package-private abstract class. There are two implementations of this class, one for pre-spark-3.3.0 and the other for Spark3.3.0. Major difference between the two is the Spark-3.3.0 version uses `ParquetColumn` to read the parquet file while the other uses `VectorizedColumnReaders` 

fixes #5257 
fixes #5357
fixes #5429
 
Signed-off-by: Raza Jafri <rjafri@nvidia.com>

### CODE REVIEW

File: sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala
Comment (razajafri) [reply sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala:44]: What is the advantage of using `MethodUtils`? 
It doesn't reduce the numbers of line of code
```
val method = MethodUtils.getMatchingMethod(classOf[VectorizedColumnReader], "readBatch", Integer.TYPE,  classOf[WritableColumnVector])
method.setAccessible(true)
```
Nor do we run in JRE prior to 1.4 where `MethodUtils` has a work around.

IMO this would bring a 3rd party in to the picture when it's not really adding any value. Please feel free to elaborate on why you made the suggestion.

File: sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala
Comment (razajafri) [reply sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala:98]: You are right

File: sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala
Comment (gerashegalov) [reply sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala]: not a big deal but there are invokeMethod flavors where you can just pass forceAccess = true

**Comment by razajafri** [comment]
@gerashegalov do you have any more suggestions?

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala
Comment (razajafri) [reply sql-plugin/src/main/311until330-nondb/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala]: That will look up the method every time as seen [here](https://github.com/apache/commons-lang/blob/80889fc45cbfb3da0a916e0fc0d51663bb970c2a/src/main/java/org/apache/commons/lang3/reflect/MethodUtils.java#L205). I don't think that's what we want

**Comment by nartal1** [comment]
I am bit confused on the description of the PR. It's mentioned as it fixes this issue: https://github.com/NVIDIA/spark-rapids/issues/5257. Here we are throwing exception that `Complex types are not supported`  for both Spark-3.3 and prior versions. There was some discussion on that issue to support for complex type. Could you please clarify if we should close the issue once this PR is merged OR keep it open for future work.

**Comment by razajafri** [comment]
> I am bit confused on the description of the PR. It's mentioned as it fixes this issue: #5257. Here we are throwing exception that `Complex types are not supported` for both Spark-3.3 and prior versions. There was some discussion on that issue to support for complex type. Could you please clarify if we should close the issue once this PR is merged OR keep it open for future work.

Great catch @nartal1 I typed the wrong error message. I have fixed it PTAL

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala
Comment (nartal1) [inline sql-plugin/src/main/330+/scala/org/apache/spark/sql/execution/datasources/parquet/ShimCurrentBatchIterator.scala:199]: Nit : missing newline.

**Review by nartal1 — DISMISSED** [review]
Just one nit. Not mandatory to address. LGTM.

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
@nartal1 PTAL

**Review by nartal1 — APPROVED** [review]

**Review by gerashegalov — APPROVED** [review]
LGTM, provided we have tests exercising this logic for 3.3.0

**Comment by razajafri** [comment]
@nartal1 @gerashegalov 
I reverted the test marked to xfail. PTAL

**Review by nartal1 — APPROVED** [review]

**Comment by nartal1** [comment]
build

**Review by gerashegalov — APPROVED** [review]

**Pull request merged**
